src/EFCore.PG/Query/Internal/NpgsqlSqlTranslatingExpressionVisitor.cs

```diff
@@ -56,17 +57,38 @@ public override SqlExpression TranslateCount(Expression expression = null)
                     _sqlExpressionFactory.Function("COUNT", new[] { _sqlExpressionFactory.Fragment("*") }, typeof(long))),
                 typeof(int), _sqlExpressionFactory.FindMapping(typeof(int)));
 
-//            // In PostgreSQL SUM() doesn't return the same type as its argument for smallint, int and bigint.
-//            // Cast to get the same type.
-//            // http://www.postgresql.org/docs/current/static/functions-aggregate.html
-//            if (sqlFunctionExpression.FunctionName == "SUM")
-//        {
-//            if (sqlFunctionExpression.Type == typeof(int))
-//                Sql.Append("::INT");
-//            else if (sqlFunctionExpression.Type == typeof(short))
-//                Sql.Append("::SMALLINT");
-//            return expr;
-//        }
+        // In PostgreSQL SUM() doesn't return the same type as its argument for smallint, int and bigint.
+        // Cast to get the same type.
+        // http://www.postgresql.org/docs/current/static/functions-aggregate.html
+        public override SqlExpression TranslateSum(Expression expression)
+        {
+            if (!(expression is SqlExpression sqlExpression))
+            {
+                sqlExpression = Translate(expression);
+            }
+
+            if (sqlExpression == null)
+            {
+                throw new InvalidOperationException(CoreStrings.TranslationFailed(expression.Print()));
+            }
+
+            var inputType = sqlExpression.Type.UnwrapNullableType();
+
+            // Note that there is no Sum over short in LINQ
+            if (inputType == typeof(int))
+                return _sqlExpressionFactory.Convert(
+                    _sqlExpressionFactory.Function("SUM", new[] { sqlExpression }, typeof(long)),
+                    inputType,
+                    sqlExpression.TypeMapping);
+
+            if (inputType == typeof(long))
+                return _sqlExpressionFactory.Convert(
+                    _sqlExpressionFactory.Function("SUM", new[] { sqlExpression }, typeof(decimal)),
```

Comment:
`SUM` returns `NULL` when there are no values to aggregate. So you should use `decimal?` here and `long?` above.

Reply:
This is not how the upstream TranslateSum works: https://github.com/aspnet/EntityFrameworkCore/blob/9186d455f088c0dd051ca04112d45342b5493bdb/src/EFCore.Relational/Query/RelationalSqlTranslatingExpressionVisitor.cs#L129. I'm not totally sure of the flow (and have no time check), but the case of empty values may be taken care of elsewhere, etc. In any case, this should be investigated in the core rather than here - let's maintain thinks like they are upstream.

Reply:
Then it's hard to me to understand type mapping. I have the same question with `cardinality`. When I should use a nullable as the return type?

Reply:
I don't really have the answer, to be honest. For Sum I'm just copying the logic that I see upstream - we should definitely look deeper at what happens exactly when Sum is invoked on an empty set of values. I'm not sure exactly what difference it makes with cardinality - I'm guessing that at the very least it affects null semantics (e.g. EF Core should add `IS NOT NULL` where relevant).

Note also that aggregate functions such as Sum are a very special case and have dedicated logic, whereas cardinality is just another function we're translating.

I don't think we should block too much on this.

Reply:
Let's ask someone from the EF team who knows, For now I'm approving the changes.

Reply:
Does it mean that the type parameter means the expected CLR type and not SQL?

Reply:
Thanks for the answer @smitpatel. The question here is how a SUM of an empty set is handled - SQL-wise a NULL is returned; where exactly does it get converted to the default of type? If I compose some expression, e.g. compare that sum's result, will I also get 0?

If so, is this a general behavior for functions whose CLR return type is non-nullable, i.e. does the pipeline automatically translate nulls to default? That would mean we never have to worry about functions returning nulls.

Reply:
OK, thanks - so this seems to be specific logic for Sum. If we're translating some other random function which may return null, that means we should be returning a nullable CLR type as appropriate, right? For example, CLR array length is being translated to a PG function called `cardinality`. As usual, if you give it a null array parameter, it will return null - so that function is currently defined to return `int?` - I'm assuming that will activate the appropriate null semantics logic.

Does all this make sense?

Reply:
Thanks for clearing that up! So it actually doesn't really matter if the CLR type of DB functions is nullable or not, in the general case.